src/machine.rs

```diff
@@ -339,29 +359,39 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
 
     fn canonical_alloc_id(mem: &Memory<'mir, 'tcx, Self>, id: AllocId) -> AllocId {
         let tcx = mem.tcx;
-        // Figure out if this is an extern static, and if yes, which one.
-        let def_id = match tcx.alloc_map.lock().get(id) {
-            Some(GlobalAlloc::Static(def_id)) if tcx.is_foreign_item(def_id) => def_id,
+        let alloc = tcx.alloc_map.lock().get(id);
+        match alloc {
+            Some(GlobalAlloc::Static(def_id)) if tcx.is_foreign_item(def_id) => {
+                // Figure out if this is an extern static, and if yes, which one.
+                let attrs = tcx.get_attrs(def_id);
+                let link_name = match attr::first_attr_value_str_by_name(&attrs, sym::link_name) {
+                    Some(name) => name,
+                    None => tcx.item_name(def_id),
+                };
+                // Check if we know this one.
+                if let Some(canonical_id) = mem.extra.extern_statics.get(&link_name) {
+                    trace!("canonical_alloc_id: {:?} ({}) -> {:?}", id, link_name, canonical_id);
+                    *canonical_id
+                } else {
+                    // Return original id; `Memory::get_static_alloc` will throw an error.
+                    id
+                }
+            },
+            Some(GlobalAlloc::Static(def_id)) if tcx.has_attr(def_id, sym::thread_local) => {
```

Comment:
A static can be both foreign and thread local.

Reply:
Also you should use `tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::THREAD_LOCAL)`.